accounts-db/src/accounts_db.rs

```diff
@@ -4353,6 +4403,37 @@ impl AccountsDb {
         );
     }
 
+    pub(crate) fn zero_lamport_single_ref_found(&self, slot: Slot, offset: usize) {
+        if let Some(store) = self.storage.get_slot_storage_entry(slot) {
+            let mut zero_lamport_single_ref_offsets =
+                store.zero_lamport_single_ref_offsets.write().unwrap();
+            if zero_lamport_single_ref_offsets.insert(offset) {
+                // this wasn't previously marked as zero lamport single ref
+                if zero_lamport_single_ref_offsets.len() == store.count() {
+                    // all accounts in this storage can be dead
+                    self.accounts_index.add_uncleaned_roots([slot].into_iter());
+                    self.shrink_stats
+                        .adding_dead_slots_to_clean
+                        .fetch_add(1, Ordering::Relaxed);
+                } else {
+                    if Self::is_shrinking_productive(slot, &store)
```

Comment:
nit: we can check if slot is already in self.shrink_candidate_slot here.
if it is already in, we can skip this if shrinking_productive check () ....